File: v1_16/src/main/java/me/fullidle/ficore/ficore/v1_16/PokeNPCEntityWrapperFactoryImpl.java
package me.fullidle.ficore.ficore.v1_16;

import com.pixelmonmod.pixelmon.entities.npcs.NPCEntity;
import com.pixelmonmod.pixelmon.entities.npcs.NPCTrainer;
import lombok.val;
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.IPokeStorageWrapper;
import me.fullidle.ficore.ficore.common.api.pokemon.npc.PokeNPCEntityWrapper;
import me.fullidle.ficore.ficore.common.api.pokemon.npc.PokeNPCEntityWrapperFactory;
import me.fullidle.ficore.ficore.common.bukkit.CraftWorld;
import me.fullidle.ficore.ficore.common.bukkit.entity.CraftEntity;
import net.minecraft.world.World;
import net.minecraft.world.server.ServerWorld;
import org.bukkit.Location;
import org.bukkit.entity.Entity;

public class PokeNPCEntityWrapperFactoryImpl implements PokeNPCEntityWrapperFactory<NPCEntity> {
    public static final PokeNPCEntityWrapperFactoryImpl INSTANCE = new PokeNPCEntityWrapperFactoryImpl();

    @Override
    public PokeNPCEntityWrapper<NPCEntity> create(NPCEntity entity) {
        return new PokeNPCEntityWrapperImpl(entity);
    }

    @Override
    public boolean isPokeNPCEntity(Entity entity) {
        return CraftEntity.getHandle(entity) instanceof NPCEntity;
    }

    @Override
    public PokeNPCEntityWrapper<NPCEntity> asPokeNPCEntity(Entity entity) {
        if (isPokeNPCEntity(entity)) return new PokeNPCEntityWrapperImpl((NPCEntity) CraftEntity.getHandle(entity));
        throw new IllegalArgumentException("Entity are not Pokémon NPC entities");
    }

    @Override
    public PokeNPCEntityWrapper<NPCEntity> create(Location location) {
        val world = (ServerWorld) CraftWorld.getHandle(location.getWorld());
        val trainer = new NPCTrainer(world);
        trainer.setPos(location.getX(), location.getY(), location.getZ());
        world.addFreshEntity(trainer);
        return create(trainer);
    }

    public static class PokeNPCEntityWrapperImpl extends PokeNPCEntityWrapper<NPCEntity> {
        public PokeNPCEntityWrapperImpl(NPCEntity wrapped) {
            super(wrapped);
        }

        @Override
        public IPokeStorageWrapper<?> getStorage() {
            if (!(getOriginal() instanceof NPCTrainer)) return null;
            return new PokeStorageManager.PokeStorageWrapper(((NPCTrainer) getOriginal()).getPokemonStorage());
        }

        @Override
        public void setTexture(String texture) {
            getOriginal().setCustomSteveTexture(texture);
        }

        @Override
        public Class<NPCEntity> getType() {
            return NPCEntity.class;
        }
    }
}

File: v1_21/Cob/src/main/kotlin/me/fullidle/ficore/ficore/vCob_21/PokeNPCEntityWrapperFactoryImpl.kt
package me.fullidle.ficore.ficore.vCob_21

import com.cobblemon.mod.common.api.storage.PokemonStore
import com.cobblemon.mod.common.api.storage.StorePosition
import com.cobblemon.mod.common.entity.npc.NPCEntity
import com.cobblemon.mod.common.entity.npc.NPCPlayerModelType
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.IPokeStorageWrapper
import me.fullidle.ficore.ficore.common.api.pokemon.npc.PokeNPCEntityWrapper
import me.fullidle.ficore.ficore.common.api.pokemon.npc.PokeNPCEntityWrapperFactory
import me.fullidle.ficore.ficore.common.bukkit.entity.CraftEntity
import org.bukkit.entity.Entity
import java.net.URI

object PokeNPCEntityWrapperFactoryImpl : PokeNPCEntityWrapperFactory<NPCEntity> {
    override fun create(entity: NPCEntity): PokeNPCEntityWrapper<NPCEntity> {
        return PokeNPCEntityWrapperImpl(entity)
    }

    override fun isPokeNPCEntity(entity: Entity): Boolean {
        return CraftEntity.getHandle(entity) is NPCEntity
    }

    override fun asPokeNPCEntity(entity: Entity): PokeNPCEntityWrapper<NPCEntity> {
        if (isPokeNPCEntity(entity)) return PokeNPCEntityWrapperImpl(CraftEntity.getHandle(entity) as NPCEntity)
        throw IllegalArgumentException("Entity $entity is not a PokeNPCEntity")
    }


    class PokeNPCEntityWrapperImpl(
        npc: NPCEntity
    ) : PokeNPCEntityWrapper<NPCEntity>(npc) {
        override fun getType(): Class<NPCEntity> {
            return NPCEntity::class.java
        }

        override fun getStorage(): IPokeStorageWrapper<*>? {
            return original.party?.let {
                PokeStorageManager.PokeStorageWrapper(it as PokemonStore<StorePosition>)
            }
        }

        override fun setTexture(texture: String) {
            original.loadTexture(URI.create(texture), NPCPlayerModelType.DEFAULT)
        }
    }
}
